Add command to save PAN import data

leonardola opened the pull request:
This PR adds a command to update payment methods from shopify_payments to WooPayments using a mapping file generated by stripe

**Test instructions**

If this server PR has not been merged to trunk yet make sure to checkout to that branch first
https://github.com/Automattic/woocommerce-payments-server/pull/4159

- Checkout to this PR's branch

**Creating a stripe customer and card:**
- Create a stripe customer with a card attached to it by placing an order and checking the option to save the card during checkout. Use the card `4242 4242 4242 4242`
- Take note of the token `pm_`  from `wp_woocommerce_payment_tokens` for this user
- Remove the entries for this user from `wp_woocommerce_payment_tokens` table
- Take note of the `wp__wcpay_customer_id_test` from `wp_usermeta` for this user
- Remove the `wp__wcpay_customer_id_test` entry from `wp_usermeta` table


**Mocking an Shopify Payments order**
Execute this steps if you don't have a Shopify Payments order and skip the **Importing an order** steps
- Create an order with a subscription product for the user created in the **Creating a stripe customer and card:** steps
- If a subscription was created delete it but keep the order
- Add `_order_number` to the `wp_post_meta` table and set the value to `11664` for this order
- Update the order creation date via the /wp-admin order edit page to `2023-01-07`
- Add `_original_payment_last_4` to the `post_meta` and set the value to `4242`
- Add `_original_payment_method_id` to the `post_meta` and set the value to `pm_mock`
- Add `_original_payment_gateway` to the `post_meta` and set the value to `shopify_payments`

**Importing an order**
- You won't be able to test the order import part if you don't have a Shopify Payments order.
- Take note of the customer email you created in the **Creating a stripe customer and card** steps
- Change the customer email to something else via wp-admin
- Execute: `wp migrator orders --ids="<shopify_order_id>`
- Make sure to change the `shopify_order_id` to an order that was placed using shopify payments
- Make sure to enable subscriptions for the product in the order in case it's not enabled yet.
- Change the `_order_number` in the `wp_post_meta` table to `11664` for this order
- Update the order creation date via the /wp-admin order edit page to `2023-01-07`
- Change the value of `_original_payment_last_4` in the `post_meta` table to `4242`
- Take note of the `_original_payment_method_id` from `post_meta` table
- Change the order user email to the one you took note in the previous steps.

**Importing payment methods**
- Execute the following command `wp migrator import_stripe_data_into_woopayments ` to import the customer data from stripe
- Make sure the new token was created in `wp_woocommerce_payment_tokens` for the order user

**Importing a Subscription **
- Download these files
[orders.json](https://github.com/woocommerce/migrator-cli/files/13429838/orders.json)
[subscriptions.json](https://github.com/woocommerce/migrator-cli/files/13430581/subscriptions.json)

- Run `wp migrator skio_subscriptions --subscriptions_export_file=<file_path>  --orders_export_file=<file_path>`
- Make sure the `_original_payment_gateway`, `_original_payment_method_id` and `_original_payment_last_4` in the `post_meta` table matches the ones in the order imported/created before

**Updating the payment methods in the subscription and order**
- Download the following file:
[migration.csv](https://github.com/woocommerce/migrator-cli/files/13430002/migration.csv)
Change the `<source_id_old>` to match the `_original_payment_method_id`, the `<customer_id_new>` to match the stripe customer id `cus_` you took note earlier and the `<payment_method_new>` to match the stripe payment method id `pm_` you took note earlier
- Execute the following command: 
`wp migrator add_woopayments_migration_data --migration_file=<migration_file_abs_path>`
- Now make sure `Payment Method:` is set to `Credit card / Debit card` and `Saved payment method:` is set to the correct card  in the billing data for the subscription
- Trigger a subscription renewal via the subscription actions
- Make sure the new order was charged correctly

Commit:
Added old stripe payment method id to the orders

Commit:
started to work on paypal

Commit:
added function to process payment method

Commit:
Fixed colorized texts

Commit:
Added command to save info about the PAN import from stripe

Commit:
revert removal of *

Commit:
reverted paypal changes

Commit:
moved payment method meta keys to a const

Commit:
phpcs fixes

Commit:
Merge branch 'trunk' into add/pull-old-payment-methods

# Conflicts:
#	README.md
#	includes/class-migrator-cli-payment-methods.php
#	includes/class-migrator-cli.php

Commit:
change $this to self

Commit:
fixed color reset and type comparison

Commit:
Added comments and fixed cli out colors

Commit:
added comments

Commit:
added check to prevent using a failed transaction. Added comments.

Commit:
Renamed variable

Commit:
added more cli output

nagpai commented:
🔴  - Manual test - stalled while running `wp migrator orders --ids="1007"`  with our test website credentials active in migrator. It could be because the transactions were all test.  Is this expected to run on a test website, @leonardola ?

nagpai commented on includes/class-migrator-cli-payment-methods.php:
```diff
@@ -68,12 +78,230 @@ private function search_payment_token_by_stripe_id( $saved_payment_tokens, $stri
+	 * @param WC_Order|WC_Subscription $order the order or subscription to be updated
+	 * @throws WC_Data_Exception
+	 */
+	private function process_shopify_payments( WC_Order|WC_Subscription $order ) {
```
I got a critical error on my test site running PHP 7.4. I looked up and found that the union type declaration within the function param would work in PHP 8.0+ . It may cause backward compatibility issue with sites running versions less than PHP 8.0.

We could add an `if` statement to check for Order / Subscription. 

```suggestion
	private function process_shopify_payments( $order ) {
	    if ( ! $order instanceof WC_Order || ! $order instanceof WC_Subscription ) {
			 WP_CLI::line( WP_CLI::colorize( '%RInvalid order type%n' ) );
			 return;
		 }
          .....
          ....
         }
```

leonardola replied:
I only added that for IDE hints. So I'll just remove it. A subscription is a WC_Order too so it should be ok.

nagpai replied:
That makes sense. Thanks for addressing this 👍

nagpai commented on includes/class-migrator-cli-orders.php:
```diff
@@ -622,4 +622,47 @@ private function set_line_item_taxes( $line_item, $shopify_line_item ) {
+			return;
+		}
+
+		switch ( $transaction->gateway ) {
```
Is there a reason we are using `switch .. case` instead of `if ... else here`? I see only `shopify_payments` and `default`. Are we expecting more additions in the future?

leonardola replied:
Yeah we're expecting more payment gateways. There is another PR open for PayPal Express Checkout now #10

nagpai reviewed (changes requested):
I had issues with manual testing, maybe because the test site does not have actual transactions processed by Shopify Pay. 

Added a comment about a line causing a critical error on a PHP 7.4 test site.

Commit:
Updated the transaction not found warning message. Added message about orders without any transactions

Commit:
Removed union type operator

leonardola commented:
> 🔴 - Manual test - stalled while running wp migrator orders --ids="1007" with our test website credentials active in migrator. It could be because the transactions were all test. Is this expected to run on a test website, @leonardola ?

The id for that order is actually 4172918816901. The id can be found in the url. It's not the number in the order page.

> I had issues with manual testing, maybe because the test site does not have actual transactions processed by Shopify Pay.

The Shopify Pay transaction is required to test the order import part. If you don't have a Shopify Pay transaction the transaction import will would show this error: `Capture transaction not found`. I've updated it to this warning instead `Warning: Capture transaction not found. Not going to import the transaction data for this order`.

As there are no shopify payments orders in the test site I've added the **Mocking an Shopify Payments order** steps in the PR description.

nagpai commented:
🔴  Manual testing - After mocking the shopify order, when i run the subscriptions import i seem to get this error 

```
Adding Subscription ids to orders
Processing order: 11664
Woo Order not found for Shopify Order: 11664
Creating Subscriptions
Processing subscription: 000fb1b2-8b4f-410b-a60d-2a67bd059bf1
Woo Order not found for Skio Subscription: 000fb1b2-8b4f-410b-a60d-2a67bd059bf1
```

I made sure to add the order number 11664 to post_id 63. Somehow getting this error. 

PS: sorry about the delay in this test. I had some issues with my local WC Pay setup.
